File: Data Structures & Algorithms/maximum-product-subarray/submission-2.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        res = nums[0]
        cur_max, cur_min = 1, 1
        for n in nums:
            tmp = cur_max * n
            cur_max = max(n, tmp, cur_min * n)
            cur_min = min(n, tmp, cur_min * n)
            res = max(res, cur_max)
        return res
            

File: Data Structures & Algorithms/maximum-sum-circular-subarray/submission-2.cpp
class Solution {
public:
    int maxSubarraySumCircular(vector<int>& nums) {
        int cur_max = 0, cur_min = 0, glob_max = nums[0], glob_min = nums[0], total = 0;
        for (int n : nums) {
            cur_max = max(cur_max + n, n);
            cur_min = min(cur_min + n, n);
            total+=n;
            glob_max = max(glob_max, cur_max);
            glob_min = min(glob_min, cur_min);
        }

        return glob_max < 0 ? glob_max : max(glob_max, total - glob_min);
    }
};